plugins/extractors/bigtable/middleware.go

```diff
@@ -0,0 +1,107 @@
+	"github.com/goto/meteor/utils"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type AdminClientMW struct {
+	tableDuration  metric.Int64Histogram
+	tablesDuration metric.Int64Histogram
+	projectID      string
+	instanceName   string
+	next           AdminClient
+}
+
+type InstancesAdminClientMW struct {
+	instancesDuration metric.Int64Histogram
+	projectID         string
+	next              InstanceAdminClient
+}
+
+func WithAdminClientMW(next AdminClient, projectID, instanceName string) (AdminClient, error) {
+	meter := otel.Meter("")
+
+	tablesDuration, err := meter.Int64Histogram("meteor.bigtable.client.tables.duration", metric.WithUnit("ms"))
+	if err != nil {
+		return nil, err
+	}
+
+	tableDuration, err := meter.Int64Histogram("meteor.bigtable.client.table.duration", metric.WithUnit("ms"))
+	if err != nil {
+		return nil, err
+	}
+
+	return &AdminClientMW{
+		tableDuration:  tableDuration,
+		tablesDuration: tablesDuration,
+		projectID:      projectID,
+		instanceName:   instanceName,
+		next:           next,
+	}, nil
+}
+
+func WithInstancesAdminClientMW(next InstanceAdminClient, projectID string) (InstanceAdminClient, error) {
+	meter := otel.Meter("")
+
+	instancesDuration, err := meter.Int64Histogram("meteor.bigtable.client.instances.duration", metric.WithUnit("ms"))
+	if err != nil {
+		return nil, err
+	}
+
+	return &InstancesAdminClientMW{
+		instancesDuration: instancesDuration,
+		projectID:         projectID,
+		next:              next,
+	}, nil
+}
+
+func (o *AdminClientMW) Tables(ctx context.Context) (res []string, err error) {
+	defer func(start time.Time) {
+		o.tablesDuration.Record(ctx,
+			time.Since(start).Milliseconds(),
+			metric.WithAttributes(
+				attribute.String("bt.project_id", o.projectID),
+				attribute.String("bt.error_code", utils.StatusCode(err).String()),
+				attribute.String("bt.instance_name", o.instanceName),
+			))
+	}(time.Now())
+
+	res, err = o.next.Tables(ctx)
+	return res, err
+}
+
+func (o *AdminClientMW) TableInfo(ctx context.Context, table string) (res *bigtable.TableInfo, err error) {
+	defer func(start time.Time) {
+		o.tableDuration.Record(ctx,
+			time.Since(start).Milliseconds(),
+			metric.WithAttributes(
+				attribute.String("bt.project_id", o.projectID),
+				attribute.String("bt.error_code", utils.StatusCode(err).String()),
+				attribute.String("bt.instance_name", o.instanceName),
+				attribute.String("bt.table_name", table),
+			))
+	}(time.Now())
+	res, err = o.next.TableInfo(ctx, table)
+	return res, err
+}
+
+func (o *InstancesAdminClientMW) Instances(ctx context.Context) (res []*bigtable.InstanceInfo, err error) {
+	defer func(start time.Time) {
+		o.instancesDuration.Record(ctx,
+			time.Since(start).Milliseconds(),
+			metric.WithAttributes(
+				attribute.String("bt.project_id", o.projectID),
+				attribute.String("bt.error_code", utils.StatusCode(err).String()),
+			))
+	}(time.Now())
```

Comment:
looks redundant, let's refactor it a bit

we can introduce new struct

```go
type metricRecorder struct {
	baseAttributes []attribute.KeyValue
}

func (o *metricRecorder) recordMetric(ctx context.Context, h metric.Int64Histogram, err error, attributes ...attribute.KeyValue) func(start time.Time) {
	return func(start time.Time) {
		// record here
	}
}
```

so we could setting common attributes on initialization

```go
func WithAdminClientMW(next AdminClient, projectID, instanceName string) (AdminClient, error) {
	...

	return &AdminClientMW{
		metricRecorder: metricRecorder{
			baseAttributes: []attribute.KeyValue{
				attribute.String("bt.project_id", projectID),
				attribute.String("bt.instance_name", instanceName),
			},
		},
                 ...
	}, nil
```

then the function will be simpler

```go
func (o *AdminClientMW) TableInfo(ctx context.Context, table string) (res *bigtable.TableInfo, err error) {
	defer o.recordMetric(ctx, o.tableDuration, err, attribute.String("bt.table_name", table))(time.Now())
         ...
}

Reply:
Wouldn't `err` always be `nil` in this case?

```go
defer o.recordMetric(ctx, o.tableDuration, err, attribute.String("bt.table_name", table))(time.Now())
```

Reply:
oh yea correct, can move the err to outer args

```go
defer o.recordMetric(ctx, o.tableDuration, attribute.String("bt.table_name", table))(time.Now(), err)
```

Reply:
Example in playground - https://goplay.tools/snippet/rRViR0-NDSB

Reply:
In my opinion, we can leave the code as is.

Reply:
[this](https://goplay.tools/snippet/RDjhLCLI33X ) will work 

read about [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)) and DRY principle

we potentially could reuse it on many places, not limited to this plugin

Reply:
Ya but in this case, we have subtly different instances of similar looking code. The DRY principle is fine but we shouldn't be _always_ applying it irrespective of the specific scenario. The suggested change makes it harder to follow. Sometimes keeping it simple with some repetition is better than deduplicating with a convoluted API.

Reply:
I would give higher weightage to clear code rather than the DRY principle.

Reply:
proceeded with previous approach, changing the global attributes initialization